cmake/CheckCompilerVersion.cmake

```diff
@@ -1,25 +1,47 @@
-macro(CheckCompilerVersion MIN_GCC_VERSION MIN_CLANG_VERSION MIN_MSVC_VERSION)
-
-if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
-  set(MIN_VERSION ${MIN_GCC_VERSION})
-elseif (CMAKE_C_COMPILER_ID STREQUAL "Clang")
-  set(MIN_VERSION ${MIN_CLANG_VERSION})
-elseif (CMAKE_C_COMPILER_ID STREQUAL "MSVC")
-  set(MIN_VERSION ${MIN_MSVC_VERSION})
-endif()
 
-if(NOT WIN32)
-  string(ASCII 27 ESC)
-  set(COLOR_RESET "${ESC}[m")
-  set(RED        "${ESC}[31m")
-endif()
+option(ARANGODB_SKIP_COMPILER_CHECK "Skip compiler support check.")
 
-if (NOT DEFINED MIN_VERSION)
-  message(WARNING "${RED}${CMAKE_C_COMPILER_ID}/${CMAKE_CXX_COMPILER_ID} compiler is not supported.${COLOR_RESET}")
-endif()
+macro(CheckMinVersionsSet)
+  if(NOT DEFINED ARANGODB_MIN_GCC_VERSION)
+    message(FATAL_ERROR "Please define ARANGODB_MIN_GCC_VERSION in CMakeLists.txt.")
+  endif()
+  if(NOT DEFINED ARANGODB_MIN_CLANG_VERSION)
+    message(FATAL_ERROR "Please define ARANGODB_MIN_CLANG_VERSION in CMakeLists.txt.")
+  endif()
+  if(NOT DEFINED ARANGODB_MIN_MSVC_VERSION)
+    message(FATAL_ERROR "Please define ARANGODB_MIN_MSVC_VERSION in CMakeLists.txt.")
+  endif()
+endmacro()
 
-if ((CMAKE_C_COMPILER_VERSION VERSION_LESS ${MIN_VERSION}) OR (CMAKE_CXX_COMPILER_VERSION VERSION_LESS ${MIN_VERSION}))
-  message(WARNING "${RED}${CMAKE_C_COMPILER_ID} compiler version C(${CMAKE_C_COMPILER_VERSION})/C++(${CMAKE_C_COMPILER_VERSION}) is too old. Need at least ${MIN_VERSION}.${COLOR_RESET}")
-endif()
+macro(CheckMinCompilerVersionInner MIN_VERSION)
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS ${MIN_VERSION})
+    message(
+      FATAL_ERROR
+        "ArangoDB requires at least ${CMAKE_CXX_COMPILER_ID} version ${MIN_VERSION},"
+        " detected ${CMAKE_CXX_COMPILER_VERSION}."
+        "\nYou can use ARANGODB_SKIP_COMPILER_CHECK to skip this check.")
+  endif()
+endmacro()
+
+macro(CheckCompilerVersion)
+
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    CheckMinCompilerVersionInner(${ARANGODB_MIN_GCC_VERSION})
+  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    CheckMinCompilerVersionInner(${ARANGODB_MIN_CLANG_VERSION})
+  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    CheckMinCompilerVersionInner(${ARANGODB_MIN_MSVC_VERSION})
+  else()
+    message(
+      FATAL_ERROR
+        "ArangoDB does not support CMAKE_CXX_COMPILER_ID ${CMAKE_CXX_COMPILER_ID}"
+    )
+  endif()
 
 endmacro()
+
+CheckMinVersionsSet()
```

Comment:
The original macro dealt with arguments rather than variables set in parent scope. Please stick to the original approach.

Reply:
I find it better to have some explicit name to the value under  scrutiny.

Reply:
Macro is a function and should work on its own, mixing execution scopes is not a good idea. You can always put any logic to the caller `CMakeLists.txt`.

Reply:
I am also in favor of having speaking names for variables, to improve maintainability (and that is highly required for our CMake...)

```
CheckCompilerVersion(
   11.2 # GCC
   14.0 # Clang
   19.32 # MSVC
 )
```

is something I highly dislike.

Can we go with:

```
CheckCompilerVersion(
  ARANGODB_MIN_GCC_VERSION,
 ARANGODB_MIN_CLANG_VERSION,
 ARANGODB_MIN_MSVC_VERSION
)
```

as a compromise?
Also I think it is rather verbose, and I like the approach from Markus better myself.

Reply:
Just an idea, the current approach is also fine for me: Why does it actually have to be one function / macro? We could directly write into the main CMakeLists.txt :
```
  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    CheckMinCompilerVersion(11.2)
  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
    CheckMinCompilerVersion(14.0)
  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
    CheckMinCompilerVersion(19.32)
```
Then we would just need one macro.